pkgs/development/tools/analysis/ida-free/src.json

```diff
@@ -0,0 +1,30 @@
+{
+  "armv7a-darwin": {
+    "contentLength": 78445250,
+    "hash": "sha256:b77a56913d669240b53f1cfd2deec20c3f4b00cbf286bedae46172e11f06431a",
+    "lastModified": "2023-01-24T14:55:09+0000",
+    "url": "https://out7.hex-rays.com/files/arm_idafree82_mac.app.zip",
```

Comment:
i guess you will like my "hybrid downloader" snippet : D

first this will fetch headers from original url
if size and time match, fetch the original url
if the original url changed, fallback to fetch from web archive

this works best when the original url sends correct last-modified headers, which is the case here

<details>

```nix
let
  inherit (lib.importJSON ./src.json) version sha256 url mirrorUrls contentLength lastModified;
in

stdenv.mkDerivation rec {
  pname = "talon";
  inherit version;

  # the original url is dynamic
  # when the original url changed, fallback to mirror urls.
  #src = fetchurl { inherit url sha256; }; # this would fail with the original url
  # TODO move fetcher to separate file
  src = stdenvNoCC.mkDerivation {
    name = builtins.baseNameOf url;
    outputHashMode = "flat";
    outputHashAlgo = "sha256";
    outputHash = sha256;
    inherit url mirrorUrls contentLength lastModified;
    nativeBuildInputs = [
      curl
      cacert
    ];
    # test: force download from mirror
    #contentLength=$((contentLength - 1))
    #lastModified="''${lastModified}-old"
    buildCommand = ''
      echo "fetching headers from original url $url ..."
      headers=$(curl -L -I $url)
      echo "headers:"
      echo "$headers"
      # note: -1 to remove trailing \r
      contentLengthActual=$(echo "$headers" | grep "^Content-Length:" | tail -n1)
      contentLengthActual=''${contentLengthActual:16: -1}
      lastModifiedActual=$(echo "$headers" | grep "^Last-Modified:" | tail -n1)
      lastModifiedActual=''${lastModifiedActual:15: -1}

      if [[ "$contentLengthActual" == "$contentLength" && "$lastModifiedActual" == "$lastModified" ]]
      then
        echo "fetching file from original url $url ..."
        curl -L -o $out $url
      else
        echo "note: original url has changed:"
        if [[ "$contentLengthActual" != "$contentLength" ]]; then
          echo "-Content-Length: $contentLength"
          echo "+Content-Length: $contentLengthActual"
        fi
        if [[ "$lastModifiedActual" != "$lastModified" ]]; then
          echo "-Last-Modified: $lastModified"
          echo "+Last-Modified: $lastModifiedActual"
        fi
        for url in "''${mirrorUrls[@]}"; do
          echo "fetching file from mirror $url ..."
          curl -L -o temp $url # TODO keep time?
          stat temp # debug
          contentLengthActual=$(stat -c%s temp)
          if [[ "$contentLengthActual" == "$contentLength" ]]; then
            mv temp $out
            break
          else
            echo "note: mirror url has changed:"
            echo "-Content-Length: $contentLength"
            echo "+Content-Length: $contentLengthActual"
          fi
        done
      fi
    '';
  };
```

src.json

```json
{
  "contentLength": 43716124,
  "lastModified": "Fri, 29 Jul 2022 21:48:55 GMT",
  "mirrorUrls": [
    "https://web.archive.org/web/20230103103331/https://talonvoice.com/dl/latest/talon-linux.tar.xz"
  ],
  "sha256": "ed12e865ea09421ebd758b49ebc1365e6739561bb2ab3957b03bfde7081cfb34",
  "url": "https://talonvoice.com/dl/latest/talon-linux.tar.xz",
  "version": "0.3.1"
}
```

update.sh

```
#! /usr/bin/env bash

# https://talonvoice.com/
# https://github.com/talonvoice/talon

urlbase='https://talonvoice.com/dl/latest/'

file_changelog='changelog.html'
file_archive='talon-linux.tar.xz'

source_json_file='src.json'

set -e

# check dependencies
for cmd in curl jq sha256sum; do
  if ! command -v $cmd >/dev/null; then
    echo "error: please install $cmd"
    exit 1
  fi
done

cd "$(dirname "$0")"

# parse old source info
old_source_json="$(< $source_json_file)"
contentLength="$(echo "$old_source_json" | jq -r .contentLength)"
lastModified="$(echo "$old_source_json" | jq -r .lastModified)"

# fetch actual source info
url="$urlbase$file_archive"
echo "fetching headers from $url ..."
headers=$(curl -L -I $url)
echo "headers:"
echo "$headers"
# note: -1 to remove trailing \r
contentLengthActual=$(echo "$headers" | grep "^Content-Length:" | tail -n1)
contentLengthActual=''${contentLengthActual:16: -1}
lastModifiedActual=$(echo "$headers" | grep "^Last-Modified:" | tail -n1)
lastModifiedActual=''${lastModifiedActual:15: -1}

if [[ "$contentLengthActual" == "$contentLength" && "$lastModifiedActual" == "$lastModified" ]]
then
    echo "done: source is up-to-date"
    exit
fi

echo "note: original url has changed:"
if [[ "$contentLengthActual" != "$contentLength" ]]; then
    echo "-Content-Length: $contentLength"
    echo "+Content-Length: $contentLengthActual"
fi
if [[ "$lastModifiedActual" != "$lastModified" ]]; then
    echo "-Last-Modified: $lastModified"
    echo "+Last-Modified: $lastModifiedActual"
fi

tempfile=$(mktemp --suffix=-$file_archive)
echo "fetching file from $url to $tempfile ..."
curl -L $url -o $tempfile
sha256Actual=$(sha256sum $tempfile)
sha256Actual=${sha256Actual:0:64}
rm $tempfile

url="$urlbase$file_changelog"
echo "fetching version from $url ..."
tempfile=$(mktemp --suffix=-$file_changelog)
curl -L $url -o $tempfile
# example: <h2>0.3.1 (Jul 28, 2022)</h2>
versionActual=$(head -n1 $tempfile | sed -E 's|^<h2>([0-9.]+).*?</h2>$|\1|')
rm $tempfile

# update source info
new_source_json="$(
    echo "$old_source_json" | jq --sort-keys --slurp '
        [.[0], {
            "contentLength": $contentLength,
            "lastModified": $lastModified,
            "mirrorUrls": []
        }] | add
    ' \
    --argjson contentLength "$contentLengthActual" \
    --arg lastModified "$lastModifiedActual"
)"

echo "new_source_json:"
echo "$new_source_json"

echo "$new_source_json" >$source_json_file

echo "todo: add mirror urls to $source_json_file"
```

</details>

Reply:
maybe this could be simpler with a smart downloader like aria2

feel free to create something like pkgs/build-support/fetchhttpmirrors/